[로또 - 웹, DB] 쉐이크반 미션 제출합니다.

vsh123 opened the pull request:
안녕하세요!
웹,DB 미션을 제출합니다!

추후 UI에 대한 CSS는 추가될 예정이지만 이번 미션의 주요 기능은 다 구현하였기 때문에 리뷰 요청을 드립니다.
또한 콘솔에 대한 피드백도 지속적으로 반영할 예정이니 많은 피드백 부탁드리겠습니다 :)

이번 리뷰도 잘 부탁드리겠습니다!

vsh123 committed:
docs : 1단계 기능 구현 목록 작성

vsh123 committed:
feat : LottoNumber 객체 생성

 1. 1~45개의 LottoNumber인스턴스 객체를 가지고 있는 클래스 생성

vsh123 committed:
feat : 로또 생성 기능 구현

vsh123 committed:
refactor : LottoNumber객체 리팩토링

vsh123 committed:
feat : 구입한 금액에 맞게 로또를 생성하는 기능 구현

vsh123 committed:
feat : 구입금액을 입력받고 로또생성하여 출력하는 기능 구현

vsh123 committed:
feat : 당첨 번호를 입력받는 기능 구현

vsh123 committed:
feat : 당첨 결과를 알려주는 Enum객체 생성

vsh123 committed:
feat : 당첨 결과를 저장하고 있는 객체 생성

vsh123 committed:
feat : 당첨 통계를 알려주는 기능 구현

vsh123 committed:
docs : 2단계 기능 구현 목록 작성

vsh123 committed:
feat : 보너스 볼을 입력받아 저장하는 기능 구현

vsh123 committed:
refactor : 메소드 이름 변경

vsh123 committed:
feat : 2등 계산 기능 및 출력 기능 추가

vsh123 committed:
docs : 3단계 기능 구현 목록 작성

vsh123 committed:
refactor : 수동과 자동을 만들어주는 Generator 분리

vsh123 committed:
refactor : WinningLotto 클래스 리팩토링

vsh123 committed:
refactor : LottoResult 리팩토링

vsh123 committed:
feat : 금액을 저장 및 검증하는 Money 클래스 생성

vsh123 committed:
feat : 수동 로또를 입력 받아 생성하는 기능 구현

vsh123 committed:
refactor : 일부 출력 메세지 리팩토링

vsh123 committed:
refactor : Main Controller 리팩토링

 1. void 메소드에 대한 재귀 호출 변경

vsh123 committed:
refactor : if문 분기 조건 변경

vsh123 committed:
refactor : exception 관련 피드백 반영

 1. exception 클래스 패키지 변경
 2. Exception 발생 위치 변경

vsh123 committed:
refactor : 조건문 피드백 반영

 1. not(!)을 붙이는 것 대신 조건문의 내용을 바꿈

vsh123 committed:
refactor : Controller 재귀 로직 변경

 1. 기존 재귀함수에서 반복문으로 변경
 2. 예외에 대한 모든 catch는 도메인에서 실시

vsh123 committed:
refactor : View 및 LottoGenerator 피드백 반영

 1. Model에서 View를 만들어주는 것이 아닌 필요한 데이터를 넘기는 방식으로 변경
 2. Auto와 Manual Generator를 하나의 클래스로 합침

vsh123 committed:
refactor : LottoResultGenerator 삭제

 1. LottoResultGenerator는 한번만 사용 되기 때문에 별도로 빼지 않고
    LottoResult클래스 안의 generate메소드로 대체

vsh123 committed:
refactor : LottoTickets 객체 수정

 1. LottoTickets 객체를 일급 컬렉션으로 변경

vsh123 committed:
refactor : Model,View 역할 분리

 1. View와 Model의 역할을 좀 더 명확하게 분리
 2. 자바 컨벤션 준수

vsh123 committed:
refactor : LottoGenerator 분리

 1. Auto와 Manual을 분리
 2. Manual에 대한 테스트 코드 작성

vsh123 committed:
refactor : 정적 팩토리 메소드 리팩토링

 1. 정적 팩토리 메소드에서 null을 리턴하지 않고
    정확한 값이 들어갔는지 확인하는 로직을 추가

vsh123 committed:
refactor : 테스트 코드 작성 및 Exception 위치 변경

vsh123 committed:
refactor : LottoResult 리팩토링

 1. LottoResultGenerator 생성

vsh123 committed:
refactor : try-catch 및 View에 대한 반복 로직 변경

 1. 기존 Model에서 실시하던 try-catch를 controller에서 실시
 2. Optional을 사용하여 반복문을 실행

vsh123 committed:
feat : 로또 구매 기능 프론트 구현

vsh123 committed:
refactor : 불필요 파일 제거

vsh123 committed:
feat : DB 연결 객체 생성

vsh123 committed:
feat : 현재 회차 가져오는 기능 구현

vsh123 committed:
feat : 로또 구매 및 DB저장 기능 구현

vsh123 committed:
feat : LottoTicketsService 추가

 1. 누락된 서비스 파일 추가

vsh123 committed:
feat : 결과 계산 기능 추가

vsh123 committed:
feat : 당첨 결과 저장 및 새로운 회차 시작 기능 구현

vsh123 committed:
feat : 당첨 결과 View 구현

vsh123 committed:
feat : 회차 선택 기능 구현

vsh123 committed:
feat : 해당 회차의 결과를 보여주는 기능 구현

vsh123 committed:
refactor : 회차 갱신 후 현재 회차 업데이트

vsh123 committed:
feat : 테스트 코드 작성

vsh123 committed:
refactor : UI 개선

vsh123 committed:
refactor : LottoResultController 리팩토링

vsh123 committed:
fix : 버그 수정

 1. 로또를 구입하지 않고 당첨번호를 입력 시 회차를 종료

vsh123 committed:
refactor : UI 개선

vsh123 committed:
refactor : 콘솔 피드백 반영

vsh123 committed:
Merge remote-tracking branch 'upstream/vsh123' into vsh123

vsh123 committed:
refactor : 콘솔 컨트롤러 리팩토링

pobiconan commented on src/main/java/lotto/DAO/DAOConnector.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Properties;
+
+public class DAOConnector {
+    public static Connection getConnection() {
```
메소드가 너무 많은 일을 하고 있는 것 같음.
메소드 분리해본다.

pobiconan commented on src/main/java/lotto/DAO/DAOConnector.java:
```diff
@@ -0,0 +1,55 @@
+        Properties properties = new Properties();
+        FileInputStream in;
+        try {
+            in = new FileInputStream("src/main/resources/database.properties");
```
설정 파일을  외부로 분리한 점 👍

pobiconan commented on src/main/java/lotto/DAO/LottoResultDAO.java:
```diff
@@ -0,0 +1,121 @@
+        con = DAOConnector.getConnection();
+    }
+
+    public static LottoResultDAO getInstance() {
```
singleton 패턴 적용 👍

pobiconan commented on src/main/java/lotto/DAO/LottoResultDAO.java:
```diff
@@ -0,0 +1,121 @@
+
+public class LottoResultDAO {
+    private static LottoResultDAO lottoResultDAO;
+    private Connection con;
```
singleton 패턴을 적용한 상태에서 이와 같이 Connection을 인스턴스 변수로 가지면 멀티 쓰레드 환경에서 문제가 될 수 있음.
왜 문제가 되는지 추후 학습할 것을 추천함. 인스턴스 변수가 아닌 로컬 변수로 구현할 것을 추천

pobiconan commented on src/main/java/lotto/DAO/LottoResultDAO.java:
```diff
@@ -0,0 +1,121 @@
+import java.util.List;
+import java.util.Map;
+
+public class LottoResultDAO {
```
DAO 쿼리 관련 부분에 많은 중복이 있다.
이 중복을 어떻게 제거하면 좋을까?

pobiconan commented on src/main/java/lotto/service/LottoResultService.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+import java.util.Map;
+
+public class LottoResultService {
```
모든 메소드가 클래스 메소드이기 때문에 생성자를 private으로 구현하거나, singleton 패턴을 적용하는 것은 어떨까?

pobiconan reviewed (changes requested):
구현 잘 했네요. 💯 
특히 db 설정을 외부 파일로 분리하고, DAO에 singleton 패턴 적용한 것이 인상적이네요.
피드백 몇 개 남겨요

vsh123 committed:
refactor : 서비스의 생서자 private로 변경

vsh123 committed:
feat : JdbcTemplate 생성

1. DAO간의 중복 해결을 위해 template 생성
2. SelectTemplate은 select시 사용
3. UpdateTemplate은 update, insert시 사용
4. Connection은 로컬 변수로 설정

vsh123 committed:
refactor : DAOConnector 메서드 분리

1. DAOConnector의 Connection의 로직을 메서드로 분리

vsh123 commented:
안녕하세요! 적어주신 피드백 반영하여 다시한번 리뷰요청드립니다!

피드백 내용을 반영하면서 다음과 같은 부분에 대해 공부하고 적용하여 봤습니다.

1. Spring에서는 jdbcTemplate을 지원하는 것으로 알고 있는데 그것과 비슷한 template을 만들어 봤습니다.
2. 서비스에서 모두 클래스메서드로 선언되었을 때 왜 생성자를 private으로 선언하는지, singleton으로 구현해야하는지 공부하여 적용해 봤습니다!

모든 코치님들께 항상 답을 주는 것이 아닌 공부할 수 있는 질문을 남겨주시는 것에 대해 감사드립니다!
보시고 더 수정할 내용 있으면 피드백 부탁드리겠습니다 :)

woowahan-pjs commented on src/main/java/lotto/DAO/LottoResultDAO.java:
```diff
@@ -0,0 +1,121 @@
+
+import lottogame.domain.Rank;
+
+import java.sql.Connection;
```
사용하지 않는 import

vsh123 committed:
refactor : 패키지 리네임

woowahan-pjs commented on src/main/java/lotto/DAO/DAOConnector.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.ArrayList;
+import java.util.Properties;
+
+public class DAOConnector {
```
클래스 이름은 UpperCamelCase로 작성합니다. 아래의 Google Java Style Guide를 참고해 주세요.
https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

woowahan-pjs commented on src/main/java/lotto/DAO/SelectJdbcTemplate.java:
```diff
@@ -0,0 +1,26 @@
+
+public abstract class SelectJdbcTemplate <T>{
+    public T executeQuery(String query, List<Integer> parameters) throws SQLException {
+        PreparedStatement pstmt = null;
```
Google Java Style Guide를 보면 지역 변수는 그것을 가장 먼저 사용하는 곳 근처에 선언하는 것이 좋다고 합니다. 아래의 글에서 이유를 한번 찾아보세요!

> Local variables are not habitually declared at the start of their containing block or block-like construct. Instead, local variables are declared close to the point they are first used (within reason), to minimize their scope. Local variable declarations typically have initializers, or are initialized immediately after declaration.

woowahan-pjs commented on src/main/java/lotto/DAO/UpdatejdbcTemplate.java:
```diff
@@ -0,0 +1,22 @@
+import java.sql.SQLException;
+import java.util.List;
+
+public class UpdatejdbcTemplate {
```
클래스 이름은 UpperCamelCase로 작성합니다. 아래의 Google Java Style Guide를 참고해 주세요.
https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

```suggestion
public class UpdateJdbcTemplate {
```

woowahan-pjs commented on src/main/java/lotto/DAO/SelectJdbcTemplate.java:
```diff
@@ -0,0 +1,26 @@
+import java.sql.SQLException;
+import java.util.List;
+
+public abstract class SelectJdbcTemplate <T>{
```
IntelliJ IDEA의 코드 자동 정렬 기능(⌥⌘L, Ctrl+Alt+L)을 사용하면 더 깔끔한 코드를 볼 수 있을 거예요.

woowahan-pjs commented on src/main/java/lottogame/domain/WinningLotto.java:
```diff
@@ -10,6 +10,14 @@ private WinningLotto(Lotto lotto, int bonusNumber) {
         this.bonusNumber = LottoNumber.Of(bonusNumber);
```
메서드 이름은 lowerCamelCase로 작성합니다. 아래의 Google Java Style Guide를 참고해 주세요.

> Method names are written in lowerCamelCase.

vsh123 committed:
refactor : 자바 컨벤션 준수

woowahan-pjs commented on src/main/java/lotto/DAO/LottoResultDAO.java:
```diff
@@ -0,0 +1,121 @@
+
+    public void updateLottoResult(int round, Map<Rank, Integer> result) {
+        UpdatejdbcTemplate template = new UpdatejdbcTemplate();
+        PreparedStatement pstmt = null;
```
지금은 사용하지 않는 코드인 것 같아요.

woowahan-pjs reviewed (changes requested):
- VALUE OBJECT는 불변이 특징이며, 비즈니스 로직을 가지고 있습니다.
- DATA TRANSFER OBJECT는 비즈니스 로직을 가지는 것이 아닌 서로 다른 영역 간의 데이터 이동을 위한 용도로 사용합니다. 외부로부터 데이터의 안전성을 확보할 수도 있습니다.

아래의 글이 도움 되면 좋겠어요.

- http://martinfowler.com/eaaCatalog/dataTransferObject.html
- http://www.martinfowler.com/eaaCatalog/valueObject.html

vsh123 committed:
refactor : 불필요 코드 제거, 컨벤션 준수

vsh123 committed:
refactor : 불필요 import 제거

vsh123 commented:
안녕하세요!
보내주신 피드백에 대해 반영하여 다시한번 리뷰요청드립니다!

오프라인 리뷰를 통해 궁금한 점에 대해 많이 알 수 있어서 정말 좋았습니다 :)

감사합니다!

woowahan-pjs reviewed (approved):
수많은 피드백을 반영하고자 대단히 고생 많으셨습니다. 💯 
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.
이만 Merge 하겠습니다.
